src/Orleans.CodeGenerator/ProxyGenerator.cs

```diff
@@ -151,6 +151,51 @@ MethodDeclarationSyntax CreateProxyMethod(ProxyMethodDescription methodDescripti
                     .Concat(_codeGenerator.LibraryTypes.StaticCopiers)
                     .ToList();
 
+            // Ensure to hook up the cancellation token if the method has one
+            var cancellationTokenParameter = methodSymbol.Parameters.SingleOrDefault(parameter => SymbolEqualityComparer.Default.Equals(LibraryTypes.CancellationToken, parameter.Type));
+            if (cancellationTokenParameter is not null)
+            {
+                // Throw aggressively if cancellation is already requested
+                statements.Add(
+                    ExpressionStatement(
+                        InvocationExpression(
+                            IdentifierName($"arg{cancellationTokenParameter.Ordinal}").Member("ThrowIfCancellationRequested"),
+                            ArgumentList()
+                        )
+                    )
+                );
+
+                // Register for cancellation
+                statements.Add(
+                    ExpressionStatement(
+                        InvocationExpression(
```

Comment:
I wonder if this should be a `using` expression so that the registration is disposed (so as not to leak), and the presence of a `CancellationToken` should force the method to be `async` (see line 287: https://github.com/dotnet/orleans/pull/9127/files#diff-1685cd0d6144d179427808f9771dcfe4e296a028fc1ed6dd648c6b45741198fbR287).

Reply:
`CallbackData` is now responsible for registering for cancellation and disposing of the registration. We only register when `CanBeCanceled` is true.